File: js/agentbot/agentbot.js
/**
 */
'use strict'

// to avoid issue with the self signed certificate
// process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

const https = require('https');
const base64 = require('base-64');
const crypto = require("crypto");
const axe = require('axios');
var config = require('./conf_agent.json');
		 
// Options to be used by request 
var axconfig = {
		hostname: config.baseUrl,
        port: config.port,
//        baseUrl: "https://" + config.baseUrl,
		url: "https://prod-eu.sapcctr.com/standarddemo/ecfs/authentication",
		method: "POST",
		headers: {
		    "Content-Type": "application/x-www-form-urlencoded",
		    "Authorization": "ECFAUTH"
        }
};


var userGUID, sessionCookie;
var pattern = /JSESSIONID=[0-9A-Fa-f]*/;

var subscribeReq = {
			"client_req_id": crypto.randomBytes(16).toString("hex"),
			"method": "SUBSCRIBE",
            "uri": "/users/me/properties"
		}

var subscribeQueueReq = {
		"client_req_id": crypto.randomBytes(16).toString("hex"),
		"method": "SUBSCRIBE",
        "uri": "/users/me/queues"
	}

var subscribeInteractionsReq = {
		"client_req_id": crypto.randomBytes(16).toString("hex"),
		"method": "SUBSCRIBE",
		"uri": "/users/me/interactions"
    }
    
var putGetReadyReq = {
    "client_req_id": crypto.randomBytes(16).toString("hex"),
    "method": "PUT",
    "uri": "/users/me/properties",
    "body": {work_status: "ready"}   
}

function transferChat(clientReqId, interactionGUID, destQueue) {
	return {
		"client_req_id": clientReqId,
	  	"method": "PUT",
	  	"uri": "/users/me/interactions/"+interactionGUID,
	  	"body": {	
	  		"channel_status": "transferring",
	  		"destination": destQueue
	  	}		
	}
}

function postInteraction(clientReqId) {
	return {
		"client_req_id": clientReqId,
	  	"method": "POST",
	  	"uri": "/users/me/interactions",
	  	"body": {
	  			channel_type: "chat", 
	  			channel_sub_type: "text", 
	  			destination: config.destination
	  		}	
	  	}
}

function putProperty(clientReqId) {
	return {
		"client_req_id": clientReqId,
	  	"method": "PUT",
	  	"uri": "/users/me/properties",
	  	"body": {	
	  		chat_address: config.originator, 
	  		alias: config.alias
	  	}		
	}
}

// /users/me/queues?
function subscribeQueues(clientReqId) {
	return {
		"client_req_id": clientReqId,
	  	"method": "SUBSCRIBE",
	  	"uri": "/users/me/queues",
	  	"body": {}		
	}
}



function messageInteraction(clientReqId, interactionGUID, message, originator) {
	return {
		"client_req_id": clientReqId,
	  	"method": "POST",
	  	"uri": "/users/me/interactions/"+interactionGUID+"/transcript/messages",
	  	"body": {	
	  		"message": message,
	  		"originator": config.originator
	  	}		
	}
}

var base64credentials = Buffer.from(config.username + ":" + config.password).toString('base64')
var authformdata = "Authorization=Basic%20" + base64credentials
var authUrl = "https://" + config.baseUrl + config.authUri
axe.post(authUrl, authformdata,  axconfig)
    .then(function (response) {
        console.log("Authentication response: ")
        console.log(response.data);
        console.log(response.status);
        console.log(response.statusText);
        var body = response.data;
        var data = response.data;
        
        //console.log(response.headers);
        //console.log(response.config);

        // user GUID
		userGUID = data.id;
		console.log('AUTHENTICATED USER: ' + userGUID);		
		// get the JSESSIONID cookie (used to idetify the session on the server side)
		sessionCookie = pattern.exec(response.headers['set-cookie'])[0]
		console.log('HTTP SESSION: ' + sessionCookie);
      
		if (sessionCookie) {
			const WebSocket = require('ws');
			// Establish websocket connection			
			const ws = new WebSocket(config.webSocketUrl, {
				origin: config.origin,
				protocolVersion: config.protocolVersion,				
				headers: {
					'Content-Type': 'application/json',
					'Cookie': sessionCookie					
				}
			});
		    
    	  ws.on('unexpected-response', function unexpResponse(request, response) {
    		  console.log(request);
    		  console.log(response);
    	  });
		
		
    	  ws.on('open', function open() {    		  
    		  console.log('WS CONNECTION ESTABLISHED: ' + config.baseUrl);
    		  console.log('');
    		  //ws.send(JSON.stringify(putProperty(crypto.randomBytes(16).toString("hex"))));
    		  // Subscribe to resources
    		  //ws.send(JSON.stringify(subscribeReq));
    		  // Various request examples

			  let pingTimer = setInterval(() => {
				// Send ping every 50 seconds
				ws.ping();
			 }, 50000)
    	  });
		
		  ws.on('close', function close(e) {
			  // Unsubscribe from all the resources
			  //console.log(e);
			  console.log('UNSUBSCRIBE ');
			  console.log('disconnected');
		  });

		 ws.on('ping', () => {
			console.log('got ping answering pong');
			ws.pong();
		 });
		
		 ws.on('pong', () => {
			console.log('got pong');
		 });

		ws.on('message', function incoming(data, flags) {
            var jsonmsg = JSON.parse(data)

			if (jsonmsg.body && jsonmsg.body.hasOwnProperty('transactions')) {
                var interBody = jsonmsg.body;
                console.log(jsonmsg.uri)
                console.log(interBody.transactions)
				for (var key in interBody) {
					//console.log('1: '+ interBody[key]);
					for (var trans in interBody[key]) {
                    //console.log(interBody[key][trans]);
                    let currtrans = interBody[key][trans]
                    console.log(currtrans)

					    for (var res in interBody[key][trans]) {
						//console.log(interBody[key][trans].subscribed_resource)
                            if (interBody[key][trans].subscribed_resource == 'accepted') {
								var uri = interBody[key][trans].uri;
								var intId = interBody[key][trans].uri.split("/")[4]//interBody[key][trans][res].id;
								//if (intId && intId != null) {
								ws.send(JSON.stringify(messageInteraction(crypto.randomBytes(16).toString("hex"), intId, config.message)));										  
								//}
							}	  								  								  
						}
					}
				}
			}else{
                //console.log(jsonmsg);
                if (jsonmsg.sessionid){
                    console.log("AGENT TERMINAL SESSION ID " + jsonmsg.sessionid)
                    //got session id safe to send somethng
                    ws.send(JSON.stringify(subscribeReq));
                    ws.send(JSON.stringify(subscribeInteractionsReq));    		  
                    //ws.send(JSON.stringify(postInteraction(crypto.randomBytes(16).toString("hex"))));
                    ws.send(JSON.stringify(subscribeQueueReq)); 
                    ws.send(JSON.stringify(putGetReadyReq));

                }
            }  
			  /*} catch (e) {
				  console.log(e);
			  }	*/	  		 							  		  
		  });
		
		  ws.on('error', function error(e) {
			  //console.log('error: ' + e);  	  
		  });
      }



      })
    .catch(function (error) {
        console.log(error);
      });


//Authenticate user
// Callback function is used to deal with response
var callback = function(response){	
	// Continuously update stream with data
	var body = '';
	response.setEncoding('utf8');
	response.on('data', function(data) {			
		
		body += data;
            
		// user GUID
		userGUID = JSON.parse(data).id;
		console.log('AUTHENTICATED USER: ' + userGUID);		
		// get the JSESSIONID cookie (used to idetify the session on the server side)
		sessionCookie = pattern.exec(response.headers['set-cookie'])[0]
		console.log('HTTP SESSION: ' + sessionCookie);
      
		if (sessionCookie) {
			const WebSocket = require('ws');
			// Establish websocket connection			
			const ws = new WebSocket(config.webSocketUrl, {
				origin: config.origin,
				protocolVersion: config.protocolVersion,				
				headers: {
					'Content-Type': 'application/json',
					'Cookie': sessionCookie					
				}
			});
		    
    	  ws.on('unexpected-response', function unexpResponse(request, response) {
    		  console.log(request);
    		  console.log(response);
    	  });
		
		
    	  ws.on('open', function open() {    		  
    		  console.log('WS CONNECTION ESTABLISHED: ' + config.baseUrl);
    		  console.log('');
    		  //ws.send(JSON.stringify(putProperty(crypto.randomBytes(16).toString("hex"))));
    		  // Subscribe to resources
    		  ws.send(JSON.stringify(subscribeReq));
    		  // Various request examples
    		  ws.send(JSON.stringify(subscribeInteractionsReq));    		  
			  //ws.send(JSON.stringify(postInteraction(crypto.randomBytes(16).toString("hex"))));
			  ws.send(JSON.stringify(subscribeQueueReq)); 
			  let pingTimer = setInterval(() => {
				// Send ping every 50 seconds
				ws.ping();
			 }, 50000)
    	  });
		
		  ws.on('close', function close(e) {
			  // Unsubscribe from all the resources
			  //console.log(e);
			  console.log('UNSUBSCRIBE ');
			  console.log('disconnected');
		  });

		 ws.on('ping', () => {
			console.log('got ping answering pong');
			ws.pong();
		 });
		
		 ws.on('pong', () => {
			console.log('got pong');
		 });

		  ws.on('message', function incoming(data, flags) {
			  console.log(data);
			  //try{
				  if (JSON.parse(data).body && JSON.parse(data).body.hasOwnProperty('transactions')) {
                      var interBody = JSON.parse(data).body;
					  for (var key in interBody) {
						  //console.log('1: '+ interBody[key]);
						  for (var trans in interBody[key]) {
							  //console.log(interBody[key][trans]);
							  for (var res in interBody[key][trans]) {
								  //console.log(interBody[key][trans].subscribed_resource);
								  if (interBody[key][trans].subscribed_resource == 'accepted') {
									  var uri = interBody[key][trans].uri;
									  var intId = interBody[key][trans].uri.split("/")[4]//interBody[key][trans][res].id;
									  //if (intId && intId != null) {
										//  ws.send(JSON.stringify(messageInteraction(crypto.randomBytes(16).toString("hex"), intId, config.message)));										  
									  //}
								  }	  								  								  
							  }
						  }
					  }
				  }  
			  /*} catch (e) {
				  console.log(e);
			  }	*/	  		 							  		  
		  });
		
		  ws.on('error', function error(e) {
			  //console.log('error: ' + e);  	  
		  });
      }
		
   });    
   
   response.on('end', function() {
	   // Data received completely.
	   console.log(body);
   });
   
   response.on('error', function(e){
	   console.log("Request error: " + e); 
   });
}
// Make a request to the server
//var res = https.request(options, callback);
//res.end();

File: py/agentbot/agentbot.py
import asyncio
import pathlib
import ssl
import websockets
import requests
import configparser
import base64
import uuid
import json
from jsonpath_ng.ext import parser
from jsonpath_ng import jsonpath
from urllib import request, parse

# %%


def authenticate():
    uid = input("Give userid: ")
    pwd = input("Give password: ")
    basic = uid + ":" + pwd
    basic_bytes = basic.encode("ascii") 
    b64_bytes = base64.b64encode(basic_bytes) 
    b64_str = "Basic " + b64_bytes.decode("ascii")
    data = parse.urlencode({"Authorization": b64_str}).encode()
    heads = {
        "Content-Type": "application/x-www-form-urlencoded",
        "Origin": "prod-eu.sapcctr.com",
        "Authorization": "ECFAUTH"
    }

    #req =  request.Request("https://prod-eu.sapcctr.com/standarddemo/ecfs/authentication", headers=heads,  data=data)
    #request.urlopen(req)
    authurl = 'https://prod-eu.sapcctr.com/standarddemo/ecfs/authentication'
    resp = requests.post(authurl, headers=heads, data=data) 
    #print(resp.cookies)
    print (requests.utils.dict_from_cookiejar(resp.cookies))
    dictestives = requests.utils.dict_from_cookiejar(resp.cookies)
    jsessionid = dictestives["JSESSIONID"]
    jsessionstr = "JSESSIONID=" + jsessionid
    cookiehead  = [("Cookie", jsessionstr)]
    return cookiehead


# Put user properties. You should set users email address with this message
put_user_properties = {
    "client_req_id": uuid.uuid4().hex,
	"method": "PUT",
	"uri": "/users/me/properties",
	"body": {	
	  		"chat_address": "", 
	  		"alias": ""
    }
}

# Subscribe to interactions
subscribe_me = {
	"client_req_id": uuid.uuid4().hex,
	"method": "SUBSCRIBE",
	"uri": "/users/me/properties"
}

# Subscribe to my interactions
subscribe_my_interactions = {
	"client_req_id": uuid.uuid4().hex,
	"method": "SUBSCRIBE",
	"uri": "/users/me/interactions"
}

# Put to me to work
work_status_ready = {
    "client_req_id": uuid.uuid4().hex,
    "method": "PUT",
    "uri": "/users/me/properties",
    "body": {"work_status": "ready"}   
}

# Accept interaction
def accept_chat(id):
    uri = "/users/me/interactions/" + id
    return {
    	"client_req_id": uuid.uuid4().hex,
		"method": "PUT",
		"uri": uri,
		"body": {"status":"accepted"},
		"retries":0,
		"max_retries":3,
		"last_status_code":"",
		"interval":5000,
		"type":"RTC"
    }

def chat_message(id, message):
    uri = "/users/me/interactions/" + id + "/transcript/messages"
    return { 
        "client_req_id": uuid.uuid4().hex,
        "method": "POST",
        "uri": uri,
        "body": {"message": message}		
    }

def handle_interaction(id):
    uri = "/users/me/interactions/" + id
    return {
    	"client_req_id": uuid.uuid4().hex,
		"method": "PUT",
		"uri": uri,
		"body": {"status":"handled"},
		"retries":0,
		"max_retries":3,
		"last_status_code":"",
		"interval":5000,
		"type":"RTC"
    }

current_interactions = {}
my_address = None

async def get_my_props(subsd):
    print(subsd)
    in_subpath =  subsd["uri"].strip("users/me/subscriptions/").split("/")
    subs_id = in_subpath[0] 
    if (subsd["method"]=="POST"):
        print("New POST subscription", subs_id)
        global my_address
        my_address = subsd["subscribed_resource"]["chat_address"]

async def bot_todo_smthng(in_action):
    response = None
    global current_interactions
    print(in_action["uri"])    
    print("interaction id ", (in_action["uri"].strip("users/me/interactions/").split("/")[0]))

    in_subpath =  in_action["uri"].strip("users/me/interactions/").split("/")
    in_id = in_subpath[0] 
    
    if (in_action["method"]=="POST" and in_id not in current_interactions): # we have new interaction
        print("New interaction coming", in_id)
        print(in_action)
        current_interactions[in_id] = in_action
        response = accept_chat(in_id)
    elif (in_action["method"]=="POST" and len(in_subpath) > 2): # post to interaction subresource (participants,messages)
        print("New POST abt something", in_id)
        if (in_subpath[1] == "transcript"): # message transcript
            if (in_action["subscribed_resource"]["originator"] != my_address): # check that it is not my own message
                echo = "Hello from Bot you said " +  in_action["subscribed_resource"]["message"]
                response = chat_message(in_id, echo)
            if (in_subpath[1] == "participants"): # new participant added
                print(in_subpath[2], "Added to discussion")
    elif (in_action["method"]=="PUT"):
        print("PUT coming", in_id)
        # todo handle interaction states here
        if (in_action["subscribed_resource"] == "ended" and in_action["uri"].split("/")[-1] == "status"):
            print("handling interaction")
            response = handle_interaction(in_id)
    elif (in_action["method"]=="DELETE"): # interaction delete
        print("DELETE", in_id)
        if (in_id in current_interactions):
            current_interactions.pop(in_id)
            print("interaction removed")
            
    return response

async def ws_shebang(cookiehead):
    uri = "wss://prod-eu.sapcctr.com/standarddemo/ecfs/ws_endpoint/"
    ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    ssl_context.load_default_certs()

    is_connected = False
    async with websockets.connect(
        uri, 
        ssl=ssl_context,
        extra_headers=cookiehead, 
        subprotocols="chat"
    ) as websocket:
        while True:
            try:              
                if (is_connected == False):
                    message = await websocket.recv() 
                    # sessionid might be useful later especially if multiple instances used                   
                    print("Opening of connection message first message ", message) # contains {"sessionid":"<guid>","ecfsessionid":"<guid>"} 
                    # Att the beginning of connection we might want do these
                    print("subscribing to my properties")
                    await websocket.send(json.dumps(subscribe_me))
                    print("subscribing to my interactions")
                    await websocket.send(json.dumps(subscribe_my_interactions))
                    print("setting workstatus ready")
                    await websocket.send(json.dumps(work_status_ready))
                    is_connected=True    
 
                else:
                    message = await websocket.recv()
                    json_msg = json.loads(message)
                    #print(json_msg)
                    # Using jsonpath for convenience
                    jsonpath_expression = parser.parse("$.body.transactions[?(@.uri=~'/users/me/properties')]")
                    pfound = jsonpath_expression.find(json_msg)

                    if(pfound):
                        print("FOUND properties")
                        for p in pfound:
                            # note s.value contains the data
                            await get_my_props(p.value)
                            print("MY chat address after found checking props >> ", my_address)
                    # we are only intrested of transactions that relate to my interactions
                    # we should additionally check that status code = 200
                    jsonpath_expression = parser.parse("$.body.transactions[?(@.uri=~'/users/me/interactions/')]")
                    ifound = jsonpath_expression.find(json_msg)

                    if(ifound):
                        for i in ifound:
                            # note i.value contains the data
                            todo = await bot_todo_smthng(i.value)
                            if(todo):
                                await websocket.send(json.dumps(todo))

               
            except websockets.exceptions.ConnectionClosed:
                print('ConnectionClosed')
                is_connected = False
                is_running = False
                break

# HERE we have the main...
if __name__ == "__main__":
    cookiehead = authenticate()
    asyncio.get_event_loop().run_until_complete(asyncio.wait([   
        ws_shebang(cookiehead)
    ]))
# %%
